### ISSUE TITLE
Lots of (useless?) download traffic on content assist in the new Lemminx Maven POM XML editor

### ISSUE THREAD

**Issue by LorenzoBettini** [eclipse-m2e/m2e-core#441]
(copied from here https://www.eclipse.org/lists/cross-project-issues-dev/msg18745.html)

I tried, in a brand new environment, the new 2021-12 Java distribution, eager to try the new Maven POM XML.

I simply created a new Maven project with the archetype and tried to use content assist in the Maven POM XML editor tab. My system monitor showed a huge amount of downloaded stuff (a few gigabytes). I remembered I had the same problem in the past, and this should be once again https://projects.eclipse.org/proposals/eclipse-lemminx

This was followed by a huge CPU usage from Java (building some kind of index?)

After that, I only had the proposals (Maven artifacts and versions) already part of my Maven cache... nothing from Maven Central... so what was the downloaded stuff good for?

By the way, the downloaded artifacts are nowhere in my home folder: they are in /tmp (I'm on Linux), which will be wiped away after a reboot.
These are the created dirs and their disk usage (~ 6 gigs)

```
du /tmp/nexus-maven-repository
3529916	/tmp/nexus-maven-repository-index.gz8422849806479204346.dir
2357276	/tmp/nexus-maven-repository-index.gz8909712491327057893.dir
```

Please, keep in mind that I saw no network feedback in Eclipse, I just realized that by looking at my system monitor.

Now think of the users with a traffic data plan...

To make things extremely worse, I restarted Eclipse, tried content assist again, and the download starts again from scratch! Further ~3 gigs downloaded...

Am I the only one experiencing this behavior?

Luckily, I was able to install the old M2E editor, without Lemminx, to go back to the good ol' working behavior.

**Comment by mickaelistria** [comment]
LemMinX-Maven tries to download Maven Central index by default. This happens in a separate process.

**Comment by LorenzoBettini** [comment]
OK, but...

- why does it do that when it then does not use such information for code completion?
- why does it do that every time after Eclipse restart?
- why does it store the cache in a temporary folder (which will be removed on reboot) and not in the home folder?
- But, most importantly, why do you think such behavior does not break most users' experience, being enabled by default and without any configuration section? (or am I missing such a configuration?)

**Comment by mickaelistria** [comment]
> why does it do that when it then does not use such information for code completion?

LemMinX-Maven does use index in completion.

> why does it do that every time after Eclipse restart?

Index is downloaded once, and then is sometimes updated remotely, on restart. Usually the delta is minor and the amount of downloaded info is very small.

> why does it store the cache in a temporary folder (which will be removed on reboot) and not in the home folder?

That sounds more like a question for maven-indexer. LemMinX-Maven only uses the APIs, and is configured to store index in `~/.m2/repository/index` unless configured otherwise. The Indexer API may use the tmp directory in the meantime.

> But, most importantly, why do you think such behavior does not break most users' experience, being enabled by default and without any configuration section? (or am I missing such a configuration?)

No one complained so far.

**Comment by mickaelistria** [comment]
FWIW, anyone willing to improve that is welcome to work on https://github.com/eclipse/lemminx-maven/issues/232 and https://github.com/eclipse/lemminx-maven/issues/233

**Comment by LorenzoBettini** [comment]
> > why does it do that when it then does not use such information for code completion?
> 
> LemMinX-Maven does use index in completion.

It doesn't, see my original post

> 
> > why does it do that every time after Eclipse restart?
> 
> Index is downloaded once, and then is sometimes updated remotely, on restart. Usually the delta is minor and the amount of downloaded info is very small.
> 

a delta of a few gigs is small?
Sometimes? Actually, always

> > why does it store the cache in a temporary folder (which will be removed on reboot) and not in the home folder?
> 
> That sounds more like a question for maven-indexer. LemMinX-Maven only uses the APIs, and is configured to store index in `~/.m2/repository/index` unless configured otherwise. The Indexer API may use the tmp directory in the meantime.
> 

I saw that directory, but it contains very small files... what should they contain?

> > But, most importantly, why do you think such behavior does not break most users' experience, being enabled by default and without any configuration section? (or am I missing such a configuration?)
> 
> No one complained so far.

Maybe because everything takes place in the background silently? wasting gigs of data plans?

As I said in the past if that's not easily configurable or the user is notified of such a huge download, such a mechanism should not end in production as it has been done in the latest Eclipse distribution...
For sure, it should not be the default editor...

**Comment by joakime** [comment]
I actually don't see this as a overwhelming problem.

The only thing of the list presented that concerns me is the use of the system temp directory for anything that isn't temporary.
If the download uses the system temp for downloading, and then moves it into place after confirming the download is complete and whole, that's perfectly sane IMO.

The concerns about file size or download size are outdated to me. 
Anything under 10GB is viewed as normal to me.
If you had raised this concern 10 years ago, I would have agreed with you.  But in today's world.

**Comment by iloveeclipse** [comment]
> Anything under 10GB is viewed as normal to me.

Not all users have unlimited or even fast enough internet access. 
Please also multiply that by the number of m2e users and think aboit all the environmental effects.
Alone the infrastructure that need to craft all that traffic is not for free - and it is Eclipse Foundation that need to pay bills for the traffic. 
And of course the global warming, you know...

**Comment by eclipsewebmaster** [comment]
*ahem*  on the user side, downloading 10G of anything pointlessly may be no problem. It's free.

But the Maven Centrals and Eclipses out there pay for transit. All of a sudden, it's not free. There's a reason so many "free" resources change their model (BinTray, DockerHub, TravisCI).

"But in today's world, nah."

Please don't make a useless 10G download "normal"

**Comment by mickaelistria** [comment]
The download is useful.
The download is expected to happen only once.
The Maven Central infrastructure is polled, not Eclipse's one.
This is already identified as an issue and issues are reported to the language server to get rid of index stuff by default, replacing it with a "cheaper" strategy. Instead of argumentation, energy would be better spent on implemented the proposals linked in https://github.com/eclipse-m2e/m2e-core/issues/441#issuecomment-990952686 , contributions are highly welcome and would have a huge positive impact.

**Comment by guw** [comment]
@LorenzoBettini Can you check that your Eclipse installation was not sandboxed? That might explain the "tmp" folder thing *if* you are on macOS. If you are not on macOS and please ignore my comment.

**Comment by HannesWell** [comment]
In general this seems to be a long standing issue with m2e, see [Bug 403464](https://bugs.eclipse.org/bugs/show_bug.cgi?id=403464).

**Comment by LorenzoBettini** [comment]
How can it be useful if it's not used for code completion (see my original post)?

How can it be only once if it happens every time I restart eclipse (with the same workspace)?

I'm on Linux as I said in my original post.

**Comment by mickaelistria** [comment]
It *is* used for code completion. Maybe something is failing in between and makes that you find it's not used, but still the code does use index and tries to get result from it. Why you don't see them may be the same root cause as why your index goes to a tmp directory and isn't persisted. There is probably a bug hidden here.
But to be clear, I personally won't spend some time on getting index to work better, when it's clear that an approach relying on search engine requests would give better result and eliminate this particular issue by the way.

**Comment by laeubi** [comment]
I recently have worked in a project where I had to use InteliJ, they have solved it that way that index download is disabled by default, when the index is first access the user is asked to (just as far as I remember):

- download the index only once
- download the index now and update it automatically in regular intervals
- never download the index

**Comment by laeubi** [comment]
@LorenzoBettini might it be that it is only on by default with a fresh install?

If I open a pom.xml goto dependencies and hit add it tells me index downloads are disabled.

**Comment by laeubi** [comment]
Also see this long outstanding PR:
- https://github.com/eclipse-m2e/m2e-core/pull/10

**Comment by laeubi** [comment]
Related issues:
- https://github.com/eclipse-m2e/m2e-core/issues/225
- https://github.com/eclipse-m2e/m2e-core/issues/162

**Comment by LorenzoBettini** [comment]
First, I'll answer with some more experiments for @mickaelistria , then I'll answer @laeubi in a separate post

Bug or not, in any case, the critical point is that the user has no feedback of what's going on!
No message saying "I'm going to download gigs of data, are you OK with that?"
If something goes wrong (I guess the connection is not able to download all the gigs in one shot and it will start on the next time from the start?) the download will start again from the very beginning (again, that's my impression) the next time you start Eclipse and ask for completion in the editor.
I also have the impression that you should not interact much with the editor while the download and the index construction takes place (see the final part of this message).

Anyway, I decided to give it another try on Friday.
First of all, a few information about my ADSL connection (100 Mbps download / 20 Mps upload)
and my computer (a brand new LG GRAM i7, CPU i7-1165G7, 16G RAM, NVMe SSD)

I completely removed the .m2 folder,
Eclipse for Java distribution is downloaded,
fresh workspace
create a simple Maven project with the archetype (maven-archetype-simple)

open the pom.xml

inside dependencies ask for content assist and the download starts

the system monitor shows that the download goes on at full speed (for my ADSL bandwidth)

this time everything seems to follow the "happy path" (maybe in my previous attempt the download was interrupted due to some kind of network failure?) and after about 10 minutes (!) and my laptop fan going loud due to huge CPU usage, the artifacts downloaded in /tmp goes away and my ~/.m2/repository/index is populated (~ 4.5 Gigs):

```
> du ~/.m2/repository/index
4595496 /home/bettini/.m2/repository/index/repo.maven.apache.org1923549499-index
4595500 /home/bettini/.m2/repository/index
```

Please, keep in mind that in the meantime, scared of disturbing lemminx when doing this job, I did not interact with Eclipse...

OK, let's see what I have now

```
<dependencies>
		<dependency>
			<groupId>junit</groupId>
			<artifactId>junit</artifactId>
			<version>3.8.1***ASK FOR CODE COMPLETION***</version>
		</dependency>
```

I finally get the available versions!
Cool! It works?

Let's try to add another dependency, say org.mockito:mockito-core (REMEMBER, I started with an empty .m2 folder)

```
<dependencies>
		<dependency>
			<groupId>junit</groupId>
			<artifactId>junit</artifactId>
			<version>3.8.1</version>
		</dependency>
***ASK FOR CODE COMPLETION***
```

no trace of mockito proposals (even if starting writing "mock")
commons-io is proposed (it's actually cached in the .m2 folder, maybe due to the Maven standard plugins)

Let's manually specify the groupId org.mockito (manually, because no code completion helps me here).
Then,

```
		<dependency>
			<groupId>org.mockito</groupId>
			***ASK FOR CODE COMPLETION***
		</dependency>
```

org.mockito:mockito-core is proposed and after acceptance I get

```
		<dependency>
			<groupId>org.mockito</groupId>
			<groupId>org.mockito</groupId>
			<artifactId>mockito-core</artifactId>
			<version>4.1.0</version>
		</dependency>
```

so you see, the completion reinserts also the groupId, breaking the validity of the POM

Let's add <artifactId> and ask for completion

```
    <dependency>
      <groupId>org.mockito</groupId>
      <artifactId>***ASK FOR CODE COMPLETION***</artifactId>
    </dependency>
```

mockito-core is proposed and after acceptance I get

```
    <dependency>
      <groupId>org.mockito</groupId>
      <artifactId>mockito-core</artifact
       <version>4.1.0</version>Id>
    </dependency>
```

which again breaks the validity of the XML (see the broken XML elements after the insertion of the proposal)

after mockito was added (in the .m2 cache), e.g., after manually fixing the XML and saving the POM with a valid dependency of mockito

```
	<dependencies>
		<dependency>
			<groupId>junit</groupId>
			<artifactId>junit</artifactId>
			<version>4.13.2</version>
		</dependency>
***ASK FOR CODE COMPLETION***
	</dependencies>
```

only mockito-core (which has been cached as a Maven artifact in .m2) is proposed (not, e.g., mockito-all, which is NOT part of the local .m2 cache)

the same behavior can be seen with proposals in the plugins section (even in the pluginManagement)

In the meantime in the Error Log there are tons of this

```
org.eclipse.jface.text.BadLocationException
	at org.eclipse.jface.text.TreeLineTracker.fail(TreeLineTracker.java:1054)
	at org.eclipse.jface.text.TreeLineTracker.getLineInformation(TreeLineTracker.java:1163)
	at org.eclipse.jface.text.AbstractLineTracker.getLineInformation(AbstractLineTracker.java:137)
	at org.eclipse.wst.sse.core.internal.text.BasicStructuredDocument.getLineInformation(BasicStructuredDocument.java:1461)
	at org.eclipse.lsp4e.LSPEclipseUtils.toOffset(LSPEclipseUtils.java:158)
	at org.eclipse.lsp4e.outline.SymbolsLabelProvider.getMaxSeverity(SymbolsLabelProvider.java:177)
	at org.eclipse.lsp4e.outline.SymbolsLabelProvider.getImage(SymbolsLabelProvider.java:154)
	at org.eclipse.ui.internal.navigator.NavigatorContentServiceLabelProvider.findImage(NavigatorContentServiceLabelProvider.java:204)
	at org.eclipse.ui.internal.navigator.NavigatorContentServiceLabelProvider.getColumnImage(NavigatorContentServiceLabelProvider.java:112)
	at org.eclipse.ui.internal.navigator.NavigatorContentServiceLabelProvider.getImage(NavigatorContentServiceLabelProvider.java:104)
	at org.eclipse.ui.internal.navigator.NavigatorDecoratingLabelProvider$StyledLabelProviderAdapter.getImage(NavigatorDecoratingLabelProvider.java:63)
	at org.eclipse.jface.viewers.DelegatingStyledCellLabelProvider.getImage(DelegatingStyledCellLabelProvider.java:198)
	at org.eclipse.jface.viewers.DecoratingStyledCellLabelProvider.getImage(DecoratingStyledCellLabelProvider.java:171)
	at org.eclipse.jface.viewers.DelegatingStyledCellLabelProvider.update(DelegatingStyledCellLabelProvider.java:124)
	at org.eclipse.jface.viewers.DecoratingStyledCellLabelProvider.update(DecoratingStyledCellLabelProvider.java:134)
	at org.eclipse.jface.viewers.ViewerColumn.refresh(ViewerColumn.java:144)
	...
```

I did these experiments on Friday.
Today, Monday, I opened Eclipse again, open the POM, ask for completion and... the huge download starts again!
After only 3 days!
This time I interacted with the editor during the download, and in the end, I noted that the
.m2/repository/index was cleared and now I only have a few small files

```
ll ~/.m2/repository/index/repo.maven.apache.org1923549499-index
total 16K
-rw-rw-r-- 1 bettini bettini  299 dic 13 14:29 _0.cfe
-rw-rw-r-- 1 bettini bettini 1,1K dic 13 14:29 _0.cfs
-rw-rw-r-- 1 bettini bettini  365 dic 13 14:29 _0.si
-rw-rw-r-- 1 bettini bettini  136 dic 13 14:29 segments_2
-rw-rw-r-- 1 bettini bettini    0 dic 13 14:29 write.lock
```

and guess what? the code completion is broken again, and I don't even have any version completion, e.g., for JUnit, besides the version already in the .m2 folder.
The CPU is still showing lots of usages.

I had to restart the computer so that the /tmp folder was cleared.
Restarted Eclipse, open the POM, ask for completion, wait for 10 minutes for Lemminx to do its things (download and index creation), without touching anything else and I finally had the .m2/repository/index filled up again.
Anyway, code completion still works as I described above.
Thus, basically, real code completion mechanisms work only for Maven artifacts cached in the .m2 folder.
Only versions are proposed correctly using all the available versions in Maven central.
But not for artifacts. You get only proposals for artifacts you already have in the .m2 cache. Moreover, the completion does not always leave the POM in a valid state.

As I said in the original post, I have no problem reverting to the now-called Legacy POM editor.
But think of not-so-advanced Eclipse users... they might not even realize what's going on under the hood (the huge downloads and the high CPU usage).
If you care about Eclipse users' experience, this should not be the default Maven POM editor.
It's not ready yet, I'm afraid.

**Comment by LorenzoBettini** [comment]
@laeubi you're referring to this dialog, aren't you?

![image](https://user-images.githubusercontent.com/1202254/145830704-0b09cf03-123c-41e3-b9e3-6890478ca0de.png)

As far as I know, this has nothing to do with Lemminx, which I seem to understand only implements the POM XML editor, it's the old form editor that has always been part of M2E. It refers to this preference (see the mouse pointer)

![image](https://user-images.githubusercontent.com/1202254/145831148-fe3684b5-0125-46cd-8a7d-4176aaf97df8.png)

which has been unselected by default for a few years now (for good reasons, by the way!), to avoid huge downloads...
if you select it and restart Eclipse you see

![image](https://user-images.githubusercontent.com/1202254/145831652-809a3ca0-e28b-4c90-a067-252ba0fed6dc.png)

But that has nothing to do with Lemminx index, if I understand correctly. That index is built in <WORKSPACE>/.metadata/.plugins/org.eclipse.m2e.core/nexus/

**Comment by mickaelistria** [comment]
Newer snapshots of m2e do use newer snapshots of LemMinX-Maven that doesn't use index any more and instead of Central search engine (HTTP service).
